### PR TITLE
WEBC-2048 - allow dynamic batch hostname config with wildcard and dyna…

### PR DESCRIPTION
**Pull request by gotlieb**
…mic scheduler loading

@MosheMaorKaltura

### CODE REVIEW

File: api_v3/services/BatchControlService.php [hunk 1]
```diff
@@ -170,6 +184,38 @@ private function getOrCreateScheduler(KalturaScheduler $scheduler)
...
+		}
+
+		$schedulerDb = new Scheduler();
+		$schedulerDb->setLastStatus(time());
+		$schedulerDb->setName($scheduler->name);
+		$schedulerDb->setHost($scheduler->host);
+		$schedulerDb->setDescription('');
+		$schedulerDb->save();
+
+		$schedulerDb->setConfiguredId($schedulerDb->getId() + self::SCHEDULER_BASE_ID);
```
Comment (MosheMaorKaltura) [inline api_v3/services/BatchControlService.php:215]: What if this configured ID already exist? must not override existing one.

File: batch/scheduler/KSchedulerConfig.class.php [hunk 2]
```diff
@@ -503,5 +514,18 @@ protected function getCurrentConfigFilePaths()
...
+	{
+		if (!is_null($this->getId()))
+		{
+			return $this->getId();
+		}
+		$kalturaScheduler = new KalturaScheduler();
+		$kalturaScheduler->host = $hostname;
+		$kalturaScheduler->name = $hostname;
+		$scheulder = $this->kClient->batchcontrol->getOrCreateScheduler($kalturaScheduler);
+		return $scheulder->configuredId;
```
Comment (MosheMaorKaltura) [inline batch/scheduler/KSchedulerConfig.class.php:528]: Add log with name and the allocated ID

File: batch/scheduler/KSchedulerConfig.class.php [hunk 3]
```diff
@@ -113,11 +114,21 @@ public function load()
...
 
-		parent::__construct($configFileName, $hostname, true);
+		try
+		{
+			parent::__construct($configFileName, $hostname, true);
+		}
+		catch (Exception $e)
+		{
+			$hostNamePrefix = preg_replace('/\d+$/', self::HOSTNAME_WILDCARD , $hostname);
+			parent::__construct($configFileName, $hostNamePrefix, true);
```
Comment (MosheMaorKaltura) [inline batch/scheduler/KSchedulerConfig.class.php:124]: Add log here - could not find scheduler with ID , allocating

File: api_v3/services/BatchControlService.php [hunk 4]
```diff
@@ -170,6 +184,38 @@ private function getOrCreateScheduler(KalturaScheduler $scheduler)
...
+		if($schedulerDb)
+		{
+			return $schedulerDb;
+		}
+
+		$schedulerDb = new Scheduler();
+		$schedulerDb->setLastStatus(time());
+		$schedulerDb->setName($scheduler->name);
+		$schedulerDb->setHost($scheduler->host);
+		$schedulerDb->setDescription('');
```
Comment (MosheMaorKaltura) [inline api_v3/services/BatchControlService.php:212]: Add description - Dynamically allocated id

File: api_v3/services/BatchControlService.php (hunk 1 above)
Comment (gotlieb) [reply api_v3/services/BatchControlService.php:215]: see above - if you got to that point it means you are creating a new one.

File: api_v3/services/BatchControlService.php [hunk 5]
```diff
@@ -170,8 +188,84 @@ private function getOrCreateScheduler(KalturaScheduler $scheduler)
...
+		return $schedulerDb;
+	}
+
+	/**
+	 * @param $lastConfiguredId
+	 * @return int
+	 */
+	public static function calculateConfiguredId($lastConfiguredId)
+	{
+		$dc = kDataCenterMgr::getCurrentDc();
```
Comment (MosheMaorKaltura) [inline api_v3/services/BatchControlService.php:265]: Better us getCurrentDcId

File: api_v3/services/BatchControlService.php [hunk 6]
```diff
@@ -170,8 +188,84 @@ private function getOrCreateScheduler(KalturaScheduler $scheduler)
...
+		return intval(ceil($lastConfiguredId / 10) * 10 + 2 - $dc["id"]);
```
Comment (MosheMaorKaltura) [inline api_v3/services/BatchControlService.php:266]: To avoid blocking our self from future dc's . lets use 
...*10+1+dcid

**Pull request merged**